platform/solr/solr-factory/src/test/groovy/org/codice/solr/factory/impl/SolrClientAdapterAsyncSpec.groovy

```diff
@@ -0,0 +1,1253 @@
+      def createController = new FailsafeController('SolrClient Creation')
+          .onNextExecution(doThrow(exception))
+          .and().onNextExecution(doProceed())
+      def pingController = new FailsafeController('SolrClient Ping').onNextExecution(doProceed())
+
+    when: "creating an adapter"
+      def adapter = new SolrClientAdapter(CORE, creator, createController.&with, pingController.&with)
+
+    then: "the reported core should correspond to the requested one"
+      adapter.core == CORE
+
+    and:
+      "our creator should be registered with the adapter"
+      adapter.creator.is(creator)
+
+    when: "waiting for failsafe to complete"
+      createController.waitForSuccessfulCompletion()
+      pingController.waitForSuccessfulCompletion()
+
+    then: "the adapter should be available"
+      adapter.isAvailable()
+
+    and: "the underlying client should be the one we created"
+      adapter.proxiedClient.is(client)
+
+    and: "the underlying client should never be closed"
+      0 * client.close()
+
+    and: "there should be no registered initializers or listeners"
+      !adapter.hasInitializers() && !adapter.hasListeners()
+
+    and: "verify failsafe controllers"
+      createController.verify()
+      pingController.verify()
+
+    cleanup:
+      createController?.shutdown()
+      pingController?.shutdown()
+
+    where:
+      exception << [InterruptedException, InterruptedIOException]
+  }
+
+  @Unroll
+  def 'test client #test_client_how where an API call also fails to connect after the initial failure'() {
+    given: "a Solr client that fails direct pings"
+      def client = Mock(SolrClient) {
+        // in case the real client is consulted for the first API call
+        // since either we will be rejected right away if the client is not created but if it is
+        // a ping will be issued to the client
+        ping() >> pingFailResponse
+      }
+
+    and: "controllers that mocks every execution/attempts"
+      def createController = new FailsafeController('SolrClient Creation').onNextExecution {
+        waitTo('create').onlyIf(blocking_create)
+            .then().doReturn(client)
+      }
+      def pingController = new FailsafeController('SolrClient Ping').onNextExecution {
+        waitTo('ping').onlyIf(blocking_ping)
+            .then().doReturn()
+      }
+
+    when: "creating an adapter"
+      def adapter = new SolrClientAdapter(CORE, creator, createController.&with, pingController.&with)
+
+    then: "the adapter should be unavailable because it is being initialized"
+      !adapter.available
+
+    when: "calling any api methods"
+      adapter.request(request, COLLECTION)
+
+    then: "the method should not have been delegated to the client"
+      0 * client.request(request, COLLECTION)
+
+    and: "verify it throws an unavailable exception indicating we are initializing"
+      def e = thrown(UnavailableSolrException)
+
+      e.message.contains("initializing '$CORE' client")
```

Comment:
❓Do we really need to assert on the exception messages? They are not going to be displayed to the users, so not super critical if they change. It also makes the tests brittle.

Reply:
see my previous answer to this question